symbol/resolver.go

```diff
@@ -17,41 +17,318 @@
+type DupError struct {
+	object Object
+}
+
+func (e DupError) Error() string {
+	return fmt.Sprintf("%s is already defined.", e.object.String())
+}
+
+// Define data type error
+type TypeError struct {
+	target ObjectType
+	object ObjectType
+}
+
+func (e TypeError) Error() string {
+	return fmt.Sprintf("Type mismatched : %s, %s", e.target, e.object)
+}
+
+// Define not exists error
+type NoExistError struct {
+	identifier *ast.Identifier
+}
+
+func (e NoExistError) Error() string {
+	return fmt.Sprintf("Object not exists : %s", e.identifier.Name)
+}
+
+// Define scope structure
+type Scope struct {
+	store map[string]Object
+
+	child  []*Scope
+	parent *Scope
+}
+
+func (s *Scope) Get(name string) Object {
+	scope := s
+	for scope.parent != nil {
+		if obj, ok := scope.store[name]; ok {
+			return obj
+		}
+		scope = scope.parent
+	}
+
+	if obj, ok := scope.store[name]; ok {
+		return obj
+	}
+	scope = scope.parent
+
+	return nil
 }
 
 func NewResolver() *Resolver {
+	s := NewScope()
 	return &Resolver{
-		scope: NewScope(),
-		types: make(map[ast.Expression]SymbolType),
-		defs:  make(map[*ast.Identifier]Symbol),
+		scope: s,
+		types: make(map[ast.Expression]ObjectType),
+		defs:  make(map[*ast.Identifier]Object),
+		fns:   make(map[*ast.FunctionLiteral]ObjectType),
 	}
 }
 
 // typeof returns the type of expression exp,
 // or return InvalidSymbol if not found
-func (r *Resolver) typeOf(exp ast.Expression) SymbolType {
-	return ""
+func (r *Resolver) typeOf(exp ast.Expression) ObjectType {
+	if st, ok := r.types[exp]; ok {
+		return st
+	}
+	return InvalidSymbol
 }
 
 // objectOf returns the object denoted by the specified identifier
 // or return nil if not found
-func (r *Resolver) objectOf(id *ast.Identifier) Symbol {
+func (r *Resolver) objectOf(id *ast.Identifier) Object {
+	if sym, ok := r.defs[id]; ok {
+		return sym
+	}
+	return nil
+}
+
+func NewScope() *Scope {
+	return &Scope{
+		store:  make(map[string]Object),
+		parent: nil,
+	}
+}
+
+func NewEnclosedScope(p *Scope) *Scope {
+	s := &Scope{
+		store:  make(map[string]Object),
+		child:  make([]*Scope, 0),
+		parent: p,
+	}
+	p.child = append(p.child, s)
+	return s
+}
+
+func ResolveContract(c *ast.Contract) error {
+	r := NewResolver()
+	for _, f := range c.Functions {
+		if err := ResolveFunction(f, r); err != nil {
+			return err
+		}
+	}
 	return nil
 }
 
-func (r *Resolver) ResolveContract(c *ast.Contract) error {
+func ResolveFunction(f *ast.FunctionLiteral, r *Resolver) error {
+	name := f.Name.Name
+	obj := &Function{Name: name}
+
+	r.scope.store[name] = obj
+	r.fns[f] = FunctionObject
+	r.defs[f.Name] = obj
+
+	r.scope = NewEnclosedScope(r.scope)
+
+	if err := ResolveFunctionParameter(f.Parameters, r); err != nil {
+		return nil
+	}
+
+	if err := ResolveFunctionBody(f.Body, r); err != nil {
+		return err
+	}
+
+	r.scope = r.scope.parent
 	return nil
 }
+
+func ResolveFunctionParameter(p []*ast.ParameterLiteral, r *Resolver) error {
+	for _, p := range p {
```

Comment:
Variable name is duplicated!
Although there is not any errors, i think it should be changed for readability.
ex) `for _, param := params`